raft.go

```diff
@@ -520,100 +510,221 @@ func (r *Raft) leaderLoop() {
+				matchIndexes = append(matchIndexes, peer.progress.matchIndex)
 			}
+		case peer.progress.term < r.getCurrentTerm():
+			verifiedCounters = append(verifiedCounters, 0)
+			matchIndexes = append(matchIndexes, 0)
 		}
-		metrics.AddSample([]string{"raft", "leader", "lastContact"}, float32(diff/time.Millisecond))
 	}
+	verifiedCounter := quorumGeq(verifiedCounters)
+	matchIndex := quorumGeq(matchIndexes)
 
-	// Verify we can contact a quorum
-	quorum := r.quorumSize()
-	if contacted < quorum {
-		r.logger.Printf("[WARN] raft: Failed to contact quorum of nodes, stepping down")
-		r.setState(Follower)
-		metrics.IncrCounter([]string{"raft", "transition", "leader_lease_timeout"}, 1)
+	oldCommitIndex := r.getCommitIndex()
+	if matchIndex > oldCommitIndex && matchIndex >= r.leaderState.startIndex {
+		r.updateCommitIndex(oldCommitIndex, matchIndex)
 	}
-	return maxDiff
+	r.verified(verifiedCounter)
 }
 
-// quorumSize is used to return the quorum size. This must only be called on
-// the main thread.
-// TODO: revisit usage
-func (r *Raft) quorumSize() int {
-	voters := 0
-	for _, server := range r.configurations.latest.Servers {
-		if server.Suffrage == Voter {
-			voters++
+// Internal helper to calculate new commitIndex from matchIndexes,
+// whether votes form a quorum, etc.
+// TODO: add unit tests
+func quorumGeq(values []uint64) uint64 {
+	if len(values) == 0 {
+		return 0
+	}
+	sort.Sort(uint64Slice(values))
+	return values[(len(values)-1)/2]
+}
+
+func sum(values []uint64) uint64 {
+	var total uint64
+	for _, v := range values {
+		total += v
+	}
+	return total
+}
+
+// verifyLeader must be called from the main thread for safety.
+// Causes the followers to attempt an immediate heartbeat.
+func (r *Raft) verifyLeader(futures []*verifyFuture) {
+	r.verifyCounter++
+	counter := r.verifyCounter
+	r.leaderState.verifyBatches = append(r.leaderState.verifyBatches, verifyBatch{
+		start:   counter,
+		futures: futures,
+	})
+
+	r.logger.Printf("[INFO] raft: Verifying leadership for %v requests with counter %v",
+		len(futures), counter)
+
+	// Trigger immediate heartbeats
+	r.updatePeers()
+
+	// verifying a 1-server cluster should succeed right away
+	r.computeLeaderProgress()
+}
+
+// checkLeaderLease is used to check if we can contact a quorum of nodes
+// within the last leader lease interval. If not, we need to step down,
+// as we may have lost connectivity. Returns the maximum duration without
+// contact. This must only be called from the main thread.
+func (r *Raft) checkLeaderLease() {
+	servers := len(r.configurations.latest.Servers)
+	lastContacts := make([]uint64, 0, servers)
+	if hasVote(r.configurations.latest, r.localID) {
+		lastContacts = append(lastContacts, uint64(time.Now().UnixNano()))
+	}
+	for peerID, peer := range r.peers {
+		if hasVote(r.configurations.latest, peerID) {
+			lastContacts = append(lastContacts, uint64(peer.progress.lastContact.UnixNano()))
```

Comment:
Note that peer.progress.lastContact is more conservative than before. It's the time the last completed RPC was started. That cool? This should help #108 "Make leader leases actual leases".